src/util/side_metadata/global.rs

```diff
@@ -72,129 +77,252 @@ impl MappingState {
+            if mmap_size > 0 && !try_mmap_metadata(mmap_start, mmap_size).is_mappable() {
+                return false;
             }
+        } else {
+            lsize += meta_bytes_per_chunk(spec.log_min_obj_size, spec.log_num_of_bits);
+        }
+    }
+
+    if cfg!(target_pointer_width = "32") {
+        return try_map_per_chunk_metadata_space(start, size, lsize);
+    }
+
+    true
+}
+
+pub fn try_map_metadata_address_range(
+    start: Address,
+    size: usize,
+    global_metadata_spec_vec: &[SideMetadataSpec],
+    local_metadata_spec_vec: &[SideMetadataSpec],
+) -> bool {
+    info!("try_map_metadata_address_range({}, 0x{:x})", start, size);
+    debug_assert!(start.is_aligned_to(BYTES_IN_CHUNK));
+    debug_assert!(size % BYTES_IN_CHUNK == 0);
+
+    for spec in global_metadata_spec_vec {
+        // nearest page-aligned starting address
+        let mmap_start = address_to_meta_address(*spec, start).align_down(BYTES_IN_PAGE);
+        // nearest page-aligned ending address
+        let mmap_size = address_to_meta_address(*spec, start + size)
+            .align_up(BYTES_IN_PAGE)
+            .as_usize()
+            - mmap_start.as_usize();
+        if mmap_size > 0 && !try_mmap_metadata_address_range(mmap_start, mmap_size) {
             return false;
         }
-        if munmap_first_chunk.is_none() {
-            // if first chunk is newly mapped, it needs munmap on failure
-            munmap_first_chunk = Some(res.is_mapped());
+    }
+
+    let mut lsize: usize = 0;
+
+    for spec in local_metadata_spec_vec {
+        if cfg!(target_pointer_width = "64") {
+            // nearest page-aligned starting address
+            let mmap_start = address_to_meta_address(*spec, start).align_down(BYTES_IN_PAGE);
+            // nearest page-aligned ending address
+            let mmap_size = address_to_meta_address(*spec, start + size)
+                .align_up(BYTES_IN_PAGE)
+                .as_usize()
+                - mmap_start.as_usize();
+            if mmap_size > 0 && !try_mmap_metadata_address_range(mmap_start, mmap_size) {
+                return false;
+            }
+        } else {
+            lsize += meta_bytes_per_chunk(spec.log_min_obj_size, spec.log_num_of_bits);
         }
-        aligned_start += BYTES_IN_CHUNK;
+    }
+
+    if cfg!(target_pointer_width = "32") {
+        return try_map_per_chunk_metadata_space(start, size, lsize);
     }
 
     true
 }
 
-// Try to map side metadata for the chunk starting at `start`
-fn try_mmap_metadata_chunk(
-    start: Address,
-    global_per_chunk: usize,
-    local_per_chunk: usize,
-) -> MappingState {
-    trace!(
-        "try_mmap_metadata_chunk({}, 0x{:x}, 0x{:x})",
-        start,
-        global_per_chunk,
-        local_per_chunk
-    );
-    let global_meta_start = address_to_meta_chunk_addr(start);
+// Try to map side metadata for the data starting at `start` and a size of `size`
+fn try_mmap_metadata(start: Address, size: usize) -> MappingState {
```

Comment:
I thought you added a `mmap_noreserve` in `util::memory`. Why not use it?

Reply:
I strongly suggest putting all code about calling mmap to `util::memory`, those are the code where we do memory related sys calls.

Reply:
I really don't want our Rust code to call `mmap` everywhere, and check results in different ways. We should have one or more mmap functions in `util::memory` that are general enough and can be used in our Rust code.

Reply:
Right, I will move mmap related code to `util::memory` and fix the reuse issue